pulsar-metadata/src/main/java/org/apache/pulsar/metadata/bookkeeper/PulsarRegistrationClient.java

```diff
@@ -153,4 +212,56 @@ private static Set<BookieId> convertToBookieAddresses(List<String> children) {
         }
         return newBookieAddrs;
     }
+
+    @Override
+    public CompletableFuture<Versioned<BookieServiceInfo>> getBookieServiceInfo(BookieId bookieId) {
```

Comment:
I am trying to get more sense from this issue, but I don't see this method in the stack of https://github.com/apache/pulsar/issues/17913. Call you point out the call path from metadata main thread?

Reply:
here (copied from #17913) we are waiting on the CompletableFuture returned by this method.

but without this PR the CF is expected to be completed by the main metadata store thread, that is already running the caller code (BookkeeperSchemaStorage.lambda$openLedger)

```
t java.util.concurrent.CompletableFuture.get(java.base@11.0.15.0.1/CompletableFuture.java:1998)
	at org.apache.bookkeeper.common.concurrent.FutureUtils.result(FutureUtils.java:72)
	at org.apache.bookkeeper.common.concurrent.FutureUtils.result(FutureUtils.java:61)
	at org.apache.bookkeeper.client.DefaultBookieAddressResolver.resolve(DefaultBookieAddressResolver.java:43)
	at org.apache.bookkeeper.proto.PerChannelBookieClient.connect(PerChannelBookieClient.java:532)
	at org.apache.bookkeeper.proto.PerChannelBookieClient.connectIfNeededAndDoOp(PerChannelBookieClient.java:658)
	at org.apache.bookkeeper.proto.DefaultPerChannelBookieClientPool.initialize(DefaultPerChannelBookieClientPool.java:92)
	at org.apache.bookkeeper.proto.BookieClientImpl.lookupClient(BookieClientImpl.java:217)
```

it is really complicated to see because there are so many indirections.
But the problem is here in the end.
We had the very same problem in BK when initially I implemented  `getBookieServiceInfo` in the BK core code base (you can take a look there)

Reply:
I see, `getBookieServiceInfo` is called in a blocking way in `DefaultBookieAddressResolver#resolve`.

https://github.com/apache/bookkeeper/blob/c31dff9ff8dc3ec7454a62d71a9fff2af8505ee9/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/DefaultBookieAddressResolver.java#L43